lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ActiveContextManager.cs

```diff
@@ -0,0 +1,148 @@
+        }
+
+        /// <summary>
+        /// Adding an IActiveContext to the ActiveContext collection
+        /// Throw ApplicationException if the IActiveContext already exists.
+        /// </summary>
+        /// <param name="activeContext"></param>
+        internal void Add(IActiveContext activeContext)
+        {
+            if (_activeContexts.ContainsKey(activeContext.Id))
+            {
+                var msg = string.Format(CultureInfo.InvariantCulture, "The context [{0}] received already exists.", activeContext.Id);
+                Exceptions.Throw(new ApplicationException(msg), Logger);
+            }
+            _activeContexts.Add(activeContext.Id, activeContext);
+        }
+
+        /// <summary>
+        /// Remove an IActiveContext from the ActiveContext collection
+        /// Throw ApplicationException if the IActiveContext doesn't exist.
+        /// </summary>
+        /// <param name="activeContextId"></param>
+        internal void Remove(string activeContextId)
+        {
+            if (!_activeContexts.ContainsKey(activeContextId))
+            {
+                var msg = string.Format(CultureInfo.InvariantCulture, "The context [{0}] to be removed does not exist.", activeContextId);
+                Exceptions.Throw(new ApplicationException(msg), Logger);
+            }
+            _activeContexts.Remove(activeContextId);
+        }
+
+        /// <summary>
+        /// Returns the current number of IActiveContext in the ActiveContext collection
+        /// </summary>
+        internal int NumberOfActiveContexts
+        {
+            get { return _activeContexts.Count; }
+        }
+
+        /// <summary>
+        /// Given an IFailedEvaluator, remove associated IActiveContext from the collection
+        /// Throw SystemException if associated IActiveContext doesn't exist.
+        /// Throw IMRUSystemException if more than one IActiveContexts are associated with the IFailedEvaluator
+        /// as current IMRU driver assumes that there is only one context associated with the IFailedEvalutor
+        /// </summary>
+        /// <param name="value"></param>
+        internal void RemovedFailedContextInFailedEvaluator(IFailedEvaluator value)
+        {
+            if (value.FailedContexts != null && value.FailedContexts.Count > 0)
+            {
+                if (value.FailedContexts.Count == 1)
+                {
+                    var failedContextId = value.FailedContexts[0].Id;
+                    if (!_activeContexts.ContainsKey(failedContextId))
+                    {
+                        var msg = string.Format(CultureInfo.InvariantCulture,
+                            "The active context [{0}] attached in IFailedEvaluator [{1}] is not in the Active Contexts collection.",
+                            failedContextId,
+                            value.Id);
+                        Exceptions.Throw(new SystemException(msg), Logger);
```

Comment:
I will change it to IMRUSystemException for consistency. It could be caused by improper implementation of IMRU driver.